From 9243508e605dec41da0a90c0468ba37218dbbb83 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Wed, 19 Mar 2025 12:09:52 +0100
Subject: [PATCH] counter: stm32-lptimer-cnt: don't rely on regmap to manage
 the clock

The MFD core (parent) driver uses devm_regmap_init_mmio_clk() to enable
or disable the clock upon each register access through regmap API.

It's not possible when using CONFIG_PREEMPT_RT, to manage the clock
from within regmap, when clocksource driver is used. The IRQ requires
to access the register, hence calls clk_[en|dis]able. It takes a
spin_lock which isn't allowed in hard irq context (IRQF_TIMER flag):
-> stm32_clkevent_lp_irq_handler
 -> regmap_write
  -> clk_enable
   -> rt_spin_lock
It's highlighted in logs by "BUG: scheduling while atomic..." stack dumps.

This is a precursor rework to replace devm_regmap_init_mmio_clk() use
in the MFD core parent. It requires to manage the clock directly here.

There's no functional change intended here. Instead, call pm_runtime
get/put method to manage the clock, before/after accessing the
registers.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I60452992f549725a486fd889016d9b8ce11ead6c
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/445398
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Olivier MOYSAN <olivier.moysan@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/counter/stm32-lptimer-cnt.c | 58 ++++++++++++++++++++++-------
 1 file changed, 45 insertions(+), 13 deletions(-)

diff --git a/drivers/counter/stm32-lptimer-cnt.c b/drivers/counter/stm32-lptimer-cnt.c
index 5a50f1ac4390..95deb333e457 100644
--- a/drivers/counter/stm32-lptimer-cnt.c
+++ b/drivers/counter/stm32-lptimer-cnt.c
@@ -35,10 +35,18 @@ static int stm32_lptim_is_enabled(struct stm32_lptim_cnt *priv)
 	u32 val;
 	int ret;
 
+	ret = pm_runtime_resume_and_get(priv->dev);
+	if (ret)
+		return ret;
+
 	ret = regmap_read(priv->regmap, STM32_LPTIM_CR, &val);
 	if (ret)
 		return ret;
 
+	ret = pm_runtime_put_sync_suspend(priv->dev);
+	if (ret)
+		return ret;
+
 	return FIELD_GET(STM32_LPTIM_ENABLE, val);
 }
 
@@ -48,12 +56,11 @@ static int stm32_lptim_set_enable_state(struct stm32_lptim_cnt *priv,
 	int ret;
 	u32 val;
 
-	val = FIELD_PREP(STM32_LPTIM_ENABLE, enable);
-	ret = regmap_write(priv->regmap, STM32_LPTIM_CR, val);
-	if (ret)
-		return ret;
-
 	if (!enable) {
+		ret = regmap_write(priv->regmap, STM32_LPTIM_CR, 0);
+		if (ret)
+			return ret;
+
 		ret = pm_runtime_put_sync_suspend(priv->dev);
 		if (ret < 0)
 			return ret;
@@ -64,28 +71,32 @@ static int stm32_lptim_set_enable_state(struct stm32_lptim_cnt *priv,
 
 	ret = pm_runtime_resume_and_get(priv->dev);
 	if (ret)
-		goto disable_cnt;
+		return ret;
+
+	ret = regmap_write(priv->regmap, STM32_LPTIM_CR, STM32_LPTIM_ENABLE);
+	if (ret)
+		goto disable_clk;
 
 	/* LP timer must be enabled before writing CMP & ARR */
 	ret = regmap_write(priv->regmap, STM32_LPTIM_ARR, priv->ceiling);
 	if (ret)
-		goto disable_clk;
+		goto disable_cnt;
 
 	ret = regmap_write(priv->regmap, STM32_LPTIM_CMP, 0);
 	if (ret)
-		goto disable_clk;
+		goto disable_cnt;
 
 	/* ensure CMP & ARR registers are properly written */
 	ret = regmap_read_poll_timeout(priv->regmap, STM32_LPTIM_ISR, val,
 				       (val & STM32_LPTIM_CMPOK_ARROK) == STM32_LPTIM_CMPOK_ARROK,
 				       100, 1000);
 	if (ret)
-		goto disable_clk;
+		goto disable_cnt;
 
 	ret = regmap_write(priv->regmap, STM32_LPTIM_ICR,
 			   STM32_LPTIM_CMPOKCF_ARROKCF);
 	if (ret)
-		goto disable_clk;
+		goto disable_cnt;
 
 	priv->enabled = true;
 
@@ -93,10 +104,10 @@ static int stm32_lptim_set_enable_state(struct stm32_lptim_cnt *priv,
 	return regmap_update_bits(priv->regmap, STM32_LPTIM_CR,
 				  STM32_LPTIM_CNTSTRT, STM32_LPTIM_CNTSTRT);
 
-disable_clk:
-	pm_runtime_put_sync_suspend(priv->dev);
 disable_cnt:
 	regmap_write(priv->regmap, STM32_LPTIM_CR, 0);
+disable_clk:
+	pm_runtime_put_sync_suspend(priv->dev);
 
 	return ret;
 }
@@ -106,6 +117,7 @@ static int stm32_lptim_setup(struct stm32_lptim_cnt *priv, int enable)
 	u32 mask = STM32_LPTIM_ENC | STM32_LPTIM_COUNTMODE |
 		   STM32_LPTIM_CKPOL | STM32_LPTIM_PRESC;
 	u32 val;
+	int ret;
 
 	/* Setup LP timer encoder/counter and polarity, without prescaler */
 	if (priv->quadrature_mode)
@@ -114,7 +126,17 @@ static int stm32_lptim_setup(struct stm32_lptim_cnt *priv, int enable)
 		val = enable ? STM32_LPTIM_COUNTMODE : 0;
 	val |= FIELD_PREP(STM32_LPTIM_CKPOL, enable ? priv->polarity : 0);
 
-	return regmap_update_bits(priv->regmap, STM32_LPTIM_CFGR, mask, val);
+	ret = pm_runtime_resume_and_get(priv->dev);
+	if (ret)
+		return ret;
+
+	ret = regmap_update_bits(priv->regmap, STM32_LPTIM_CFGR, mask, val);
+	if (ret) {
+		pm_runtime_put_sync_suspend(priv->dev);
+		return ret;
+	}
+
+	return pm_runtime_put_sync_suspend(priv->dev);
 }
 
 /*
@@ -154,7 +176,17 @@ static int stm32_lptim_cnt_read(struct counter_device *counter,
 	u32 cnt;
 	int ret;
 
+	ret = pm_runtime_resume_and_get(priv->dev);
+	if (ret)
+		return ret;
+
 	ret = regmap_read(priv->regmap, STM32_LPTIM_CNT, &cnt);
+	if (ret) {
+		pm_runtime_put_sync_suspend(priv->dev);
+		return ret;
+	}
+
+	ret = pm_runtime_put_sync_suspend(priv->dev);
 	if (ret)
 		return ret;
 
-- 
2.50.1

